crates/amaru-kernel/src/protocol_parameters.rs

```diff
@@ -22,38 +31,393 @@ pub struct ProtocolParameters {
     pub ref_script_cost_stride: u32,
     pub ref_script_cost_multiplier: RationalNumber,
 
-    pub max_epoch: Epoch,
-    pub pool_influence: RationalNumber,
-    pub desired_pool_count: u32,
-    pub treasury_expansion_rate: RationalNumber,
-    pub monetary_expansion_rate: RationalNumber,
-    pub collateral_percentage: u32,
+    // Technical group
+    pub max_epoch: EpochInterval,
+    pub optimal_stake_pools_count: u16,
+    pub pledge_influence: RationalNumber,
+    pub collateral_percentage: u16,
     pub cost_models: CostModels,
 
+    // Governance group
     pub pool_thresholds: PoolThresholds,
     pub drep_thresholds: DrepThresholds,
-    pub cc_min_size: u32,
-    pub cc_max_term_length: u32,
-    pub gov_action_lifetime: u32,
+    pub cc_min_size: u16,
+    pub cc_max_term_length: EpochInterval,
+    pub gov_action_lifetime: EpochInterval,
     pub gov_action_deposit: Coin,
     pub drep_deposit: Coin,
-    pub drep_activity: Epoch,
+    pub drep_expiry: Epoch,
+}
+
+fn decode_u16(d: &mut Decoder<'_>) -> Result<u16, cbor::decode::Error> {
+    #[allow(clippy::wildcard_enum_match_arm)]
+    match d.datatype()? {
+        cbor::data::Type::U8 => Ok(d.u8()? as u16),
+        cbor::data::Type::U16 => Ok(d.u16()?),
+        _ => Err(cbor::decode::Error::message("Expected u8/u16 type")),
+    }
 }
```

Comment:
Not so sure about this. I believe I understand the rationale, but we need to ensure that the decoding behavior is identical to the Haskell node. If the Haskell node fails to decode u8 as u16, we should to.

Reply:
Apparently CBOR already support using smaller value natively and most decoder (including our own) handle the upcast at read time natively. So this code was useless :)